src/interpolation/ArborX_InterpMovingLeastSquares.hpp

```diff
@@ -0,0 +1,263 @@
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_INTERP_MOVING_LEAST_SQUARES_HPP
+#define ARBORX_INTERP_MOVING_LEAST_SQUARES_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_DetailsLegacy.hpp>
+#include <ArborX_GeometryTraits.hpp>
+#include <ArborX_HyperBox.hpp>
+#include <ArborX_IndexableGetter.hpp>
+#include <ArborX_InterpDetailsCompactRadialBasisFunction.hpp>
+#include <ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp>
+#include <ArborX_InterpDetailsPolynomialBasis.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Profiling_ScopedRegion.hpp>
+
+#include <optional>
+
+namespace ArborX::Interpolation::Details
+{
+
+// This is done to avoid a clash with another predicates access trait
+template <typename Points>
+struct MLSTargetPointsPredicateWrapper
+{
+  Points target_points;
+  int num_neighbors;
+};
+
+} // namespace ArborX::Interpolation::Details
+
+namespace ArborX
+{
+
+template <typename Points>
+struct AccessTraits<
+    Interpolation::Details::MLSTargetPointsPredicateWrapper<Points>,
+    PredicatesTag>
+{
+  KOKKOS_INLINE_FUNCTION static auto size(
+      Interpolation::Details::MLSTargetPointsPredicateWrapper<Points> const &tp)
+  {
+    return AccessTraits<Points, PrimitivesTag>::size(tp.target_points);
+  }
+
+  KOKKOS_INLINE_FUNCTION static auto
+  get(Interpolation::Details::MLSTargetPointsPredicateWrapper<Points> const &tp,
+      int const i)
+  {
+    return nearest(
+        AccessTraits<Points, PrimitivesTag>::get(tp.target_points, i),
+        tp.num_neighbors);
+  }
+
+  using memory_space =
+      typename AccessTraits<Points, PrimitivesTag>::memory_space;
+};
+
+} // namespace ArborX
+
+namespace ArborX::Interpolation
+{
+
+template <typename MemorySpace, typename FloatingCalculationType = double>
+class MovingLeastSquares
+{
+public:
+  template <typename ExecutionSpace, typename SourcePoints,
+            typename TargetPoints, typename CRBF = CRBF::Wendland<0>,
+            typename PolynomialDegree = PolynomialDegree<2>>
+  MovingLeastSquares(ExecutionSpace const &space,
+                     SourcePoints const &source_points,
+                     TargetPoints const &target_points, CRBF = {},
+                     PolynomialDegree = {},
+                     std::optional<int> num_neighbors = std::nullopt)
```

Comment:
I would still prefer
```suggestion
  template <typename ExecutionSpace, typename SourcePoints,
            typename TargetPoints, typename CRBF = CRBF::Wendland<0>,
            int degree = 2>
  MovingLeastSquares(ExecutionSpace const &space,
                     SourcePoints const &source_points,
                     TargetPoints const &target_points, CRBF = {},
                     PolynomialDegree<degree> = {},
                     std::optional<int> num_neighbors = std::nullopt)
```

Reply:
I think it's slightly less intuitive to use a raw value as a template parameter. If someone reads the implementation, they might think they can change the degree by changing the value in the template (which is not possible)